Rx.NET/Source/src/System.Reactive/Internal/WasmPlatformEnlightenmentProvider.cs

```diff
@@ -0,0 +1,60 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the Apache 2.0 License.
+// See the LICENSE file in the project root for more information. 
+
+//
+// WARNING: The full namespace-qualified type name should stay the same for the discovery in System.Reactive.Core to work!
+//
+using System.ComponentModel;
+using System.Diagnostics;
+using System.Reactive.Concurrency;
+using System.Reactive.Linq;
+using System.Reflection;
+using System.Runtime.InteropServices;
+
+namespace System.Reactive.PlatformServices
+{
+    /// <summary>
+    /// (Infrastructure) Provider for platform-specific framework enlightenments.
+    /// </summary>
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    public class WasmPlatformEnlightenmentProvider : CurrentPlatformEnlightenmentProvider
+    {
+        private readonly static bool _isWasm = RuntimeInformation.IsOSPlatform(OSPlatform.Create("WEBASSEMBLY"));
+
+        /// <summary>
+        /// (Infastructure) Tries to gets the specified service.
+        /// </summary>
+        /// <typeparam name="T">Service type.</typeparam>
+        /// <param name="args">Optional set of arguments.</param>
+        /// <returns>Service instance or <c>null</c> if not found.</returns>
+        public override T GetService<T>(object[] args) //where T : class
+        {
+            var t = typeof(T);
+
+#if !NO_THREAD || WINDOWS
+            if (t == typeof(IConcurrencyAbstractionLayer))
+            {
+#if NETSTANDARD2_0
+                if (_isWasm)
```

Comment:
@glennawatson 
I think we can get rid of `_isWasm`, because we are calling this explicitly using [`EnableWasm`](https://github.com/reactiveui/System.Reactive.Wasm/pull/1/commits/ede5563100d85e3524a386332423ddae43885817#diff-a6eb15b3179ff7f3909d42e67877d402R13), so while all of this is an external lib, `_isWasm` is redundant nevermind.

Reply:
The best way to determine if this layer should be used to to try to create a Thread. If it fails with NotSupportedException, then it should be used. Otherwise, the default from Rx can be used instead.

Reply:
You mean a non-wasm Thread? What's the proper way to test a wasm project?

Reply:
Testing for Wasm specifically is not useful, feature testing is best. Just create a new `System.Threading.Thread`, if it raises an exception, you're most probably in a Wasm environment.

Reply:
Yeah I changed it over to 

```cs
bool allowsThreads = true;

            try
            {
                new Thread(() => { }).Start();
            }
            catch (NotSupportedException)
            {
                allowsThreads = false;
            }

            if (!allowsThreads && t == typeof(IConcurrencyAbstractionLayer))
            {
                return (T)(object)new ConcurrencyAbstractionLayerWasmImpl();
            }

            if (t == typeof(IScheduler) && args != null)
            {
                return (T)(object)WasmScheduler.Default;
            }

            return base.GetService<T>(args);
```

Reply:
Should we make `allowsThreads` static, so we don't have to we do the throwing/catching/activation only once?

Reply:
Sounds like a good idea.

Reply:
[OK](https://github.com/reactiveui/System.Reactive.Wasm/pull/1/commits/2a8cdb9a1b5b0bc20e9d04c033ae5f783c08a17e).